### PR TITLE
Feature/beam package improvements

### PR DESCRIPTION
**Pull request by ericbmerritt**
###### Motivation for this change

This includes to major changes to beam packages. One is the ability to add debugging information on a global and per package basis. The second is the ability to override phases for hand written beam expressions.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

---

This change is [Reviewable](https://reviewable.io/reviews/nixos/nixpkgs/16122)

### CODE REVIEW

File: pkgs/development/beam-modules/build-mix.nix [hunk 1]
```diff
@@ -8,12 +8,20 @@
...
 , meta ? {}
+, enableDebugInfo ? false
 , ... }@attrs:
 
 with stdenv.lib;
 
 let
+  debugInfoFlag = if enableDebugInfo || elixir.debugInfo
+  then "--debug-info"
+  else "";
```
Comment (zimbatm) [inline pkgs/development/beam-modules/build-mix.nix]: you can also use `lib.optionalString (enableDebugInfo || elixir.debugInfo) "--debug-info";`

Comment (ericbmerritt) [inline pkgs/development/beam-modules/build-mix.nix]: thanks will make the changes.

**Comment by zimbatm** [comment]
Using the nox-review tool, I got some failure on the ejabberd package. Do you mind taking a look?

**Comment by zimbatm** [comment]
```
===> Upgrading erlware_commons ({pkg,<<"erlware_commons">>,<<"0.20.0">>})
===> A request is being made that violates Nix hermicity This request has been stopped. Details of the request are as follows:
===> Requesnt: get https://s3.amazonaws.com:443/s3.hex.pm/tarballs/erlware_commons-0.20.0.tar?
Makefile:39: recipe for target 'install' failed
make: *** [install] Error 1
builder for ‘/nix/store/3v0lbav9pg6q0m2wskp321jiwirfazar-hex2nix-0.0.5.drv’ failed with exit code 2
```

**Comment by ericbmerritt** [comment]
@zimbatm I just noticed the same thing and will have some fixes out shortly.

**Comment by zimbatm** [comment]
Feel free to ping me when you're happy with the PR. I don't get notified if you just push the code.

**Comment by ericbmerritt** [comment]
@zimbatm 

`ejabberd` seems to be failing because the hash to one of its deps has changed. A dep not managed by these beam packages. Its rebuilding because it does depend on elixir which we made a small change to. I could go and 'fix' the hash, but i am not the maintainer of ejabberd and have no idea what the expected behavior is. 

What is the usual way to resolve this problem?

```
output path ‘/nix/store/5hyiq0gm437day6dfb8dbdknbwd7ysg3-ejabberd-deps-16.04’ has r:sha256 hash ‘1iphn0p6gq79arsfr3dgjyypfvk0w8dvmwv9i7kxdzp0c61j8wcy’ when ‘0jrxd3v6p35wjrpw0jbb6m8nrm2lffhfgdcqmzcvsvjzji7nlkzx’ 
was expected

```

**Comment by joachifm** [comment]
cc @abbradar is one of the ejabberd maintainers, they may know more.

**Comment by abbradar** [comment]
See https://github.com/NixOS/nixpkgs/pull/16037

**Pull request merged**